src/core/namespace.c

```diff
@@ -287,10 +288,18 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
+                // mount() always follow symlinks, so we follow symlinks during the stat() as well
+                r = stat(m->path, &target);
+                if (r != 0)
+                        return r;
+                if (!S_ISDIR(target.st_mode))
+                        what = "/dev/null";
+                else {
+                        what = "/run/systemd/inaccessible";
+                }
                 break;
```

Comment:
I think it would be a good idea to keep the type of file node in tact when over-mounting it. i.e. overmount dirs with dirs, files with files, device nodes with device nodes, fifos with fifos and sockets with sockets. I'd thus propose that we introduce systematically:

* /run/systemd/inaccessible-directory as a directory of mode 000, i.e. rename from the current /run/systemd/inaccessible
* /run/systemd/inaccessible-file as a file of mode 000
* /run/systemd/inaccessible-socket as a socket of mode 000 (created with mknod(2), without anything bound to it)
* /run/systemd/inaccessible-fifo as a fifo of mode 000
* as well as /run/systemd/inaccessible-device as a copy of /dev/null of mode 000. Please do not hardcode major/minor of /dev/null to create this however as we generally try to avoid that. Instead create a new device node with mknod() based on the major/minor first read from the real /dev/null with stat().

I think it's OK to map both character and block devices to /run/systemd/inaccessible-device, i.e. both of them to a character device. I think mapping a device node to a device node is good enough.

As we cannot overmount symlinks we won't need /run/systemd/inaccessible-symlink I figure...